### PR TITLE
12-heartsteps-v1-data: Dataloader interface + heartsteps example

### LINKED ISSUE #12: HeartSteps V1 data
Add support for https://github.com/klasnja/HeartStepsV1:
 - [ ] Integrate https://github.com/calebchin/DistributionalNearestNeighbors/blob/main/experiments/structure_hs_data.ipynb into the present repo
 - [ ] Consult the group to decide whether we should save the pre-processed data in a central repository (e.g., HF)

### PR DESCRIPTION
**Pull request by calebchin**
Resolves #12 and possibly #18. Created data loader interface (see `dataloader_base.py` and `dataloader_factory.py`) and implemented the HeartSteps V1 dataset processing as an example case of subclassing this structure. Minimal working example is shown in "examples/datasets/heartsteps/test_dataloader.ipynb", where the nearest neighbor is imported as a package.

### CODE REVIEW

File: src/nearest_neighbors/dataloader_base.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+"""Abstract class for experiment/real data loading. There are two key attributes:"""
+from abc import ABC, abstractmethod
+import numpy as np
+import os
+
+class NNDataLoader(ABC):
+    # this is an abstract attribute to contain the URLs of the data of subclassers
+    urls : dict
+    supported_aggs = ['mean', 'sum', 'median', 'std', 'variance']
+    def __init__(self, download:bool=False, save_dir : str="./", agg:str="mean", save_processed:bool=False):
```
Comment (aashish-khub) [inline src/nearest_neighbors/dataloader_base.py:10]: this is pedantic, but it creates illegibility to have "download:bool", "save_dir : str", etc. Use colon and one space (e.g. "download: bool", "save_dir: str") everywhere. Follow the example @jacobf18 wrote here for style: https://github.com/aashish-khub/NearestNeighbors/blob/main/src/nearest_neighbors/nnimputer.py

Comment (aashish-khub) [reply src/nearest_neighbors/dataloader_base.py:10]: also this applies elsewhere in the code too

File: src/nearest_neighbors/datasets/heartsteps/loader.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    
+    def download_data(self) -> None:
+        """Download the data from the remote source through urls."""
+        print("Downloading HeartSteps V1 data...")
+        df_steps, df_suggestions = self._load_data()
+        df_steps.to_csv(os.path.join(self.save_dir, "jbsteps.csv"), index=False)
+        df_suggestions.to_csv(os.path.join(self.save_dir, "suggestions.csv"), index=False)
+
+
+    def process_data_scalar(self, cached:bool=False, agg:str="mean") -> tuple[np.ndarray, np.ndarray]:
```
Comment (aashish-khub) [inline src/nearest_neighbors/datasets/heartsteps/loader.py:26]: What about this method is specific to HeartSteps? Shouldn't this be in the superclass?

File: src/nearest_neighbors/datasets/heartsteps/loader.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        """Process the data into scalar setting."""
+        df_steps, df_suggestions = self._load_data(cached)
+        Data, _, Mask = self._proc_dist_data(df_steps, df_suggestions)
```
Comment (aashish-khub) [inline src/nearest_neighbors/datasets/heartsteps/loader.py:29]: why are we capitalizing Data and Mask here? Conventionally variables start with lowercase letters in python (as you have done elsewhere in this doc)

File: src/nearest_neighbors/datasets/heartsteps/loader.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+            Data = np.std(Data, axis=2)
+        elif agg == "variance":
+            Data = np.var(Data, axis=2)
+        else:
+            raise ValueError(
+                "agg must be one of 'mean', 'sum', 'median', 'std', or 'variance'"
+            )
+
+        Data = np.squeeze(Data)
+        Data = Data.astype(object)
```
Comment (aashish-khub) [inline src/nearest_neighbors/datasets/heartsteps/loader.py:46]: Why is this being done? Feel like it warrants some explanation here. I'm not convinced that this is necessary at all.

File: src/nearest_neighbors/datasets/heartsteps/loader.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+        print("Done!")
+        return Data, Mask
+
+
+    def process_data_distribution(self, cached:bool=False) -> tuple[np.ndarray, np.ndarray]:
+        """Process the data into distribution setting."""
+        df_steps, df_suggestions = self._load_data(cached)
+
+        _, Data2d, Mask = self._proc_dist_data(df_steps, df_suggestions)
+        print("Done!")
```
Comment (aashish-khub) [inline src/nearest_neighbors/datasets/heartsteps/loader.py:60]: avoid comments like these. Problem is that when you run a bunch of code you see a bunch of "Done!" or "Complete!" lines and lose context. 

Either comment these out for now, or set up a verbose_mode flag that, when `True`, prints stuff like "<timestamp>: Completed processing data distribution"

File: src/nearest_neighbors/datasets/heartsteps/loader.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        Data = Data.astype(object)
+        # write data to output_dir
+        if self.save_processed != "":
+            np.save(os.path.join(self.save_dir, "data.npy"), Data, allow_pickle=True)
+            np.save(os.path.join(self.save_dir, "mask.npy"), Mask, allow_pickle=True)
+        print("Done!")
+        return Data, Mask
+
+
+    def process_data_distribution(self, cached:bool=False) -> tuple[np.ndarray, np.ndarray]:
```
Comment (aashish-khub) [inline src/nearest_neighbors/datasets/heartsteps/loader.py:55]: Again, why is this here and not in the superclass? Seems not specific to heartsteps at all

File: src/nearest_neighbors/datasets/heartsteps/loader.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+        max_study_day: int = 52,
+        day_dec: int = 5,
+        num_measurements: int = 12,
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """Iteratively transform the processed HeartSteps data into a 4d tensor"""
+        warnings.simplefilter(action='ignore', category=pd.errors.PerformanceWarning)
+        final_M = np.zeros((users, max_study_day, day_dec, num_measurements))
+        final_A = np.zeros((users, max_study_day, day_dec))
+
+        for user in range(1, users + 1):
```
Comment (aashish-khub) [inline src/nearest_neighbors/datasets/heartsteps/loader.py:94]: This whole triple nested loop seems... a bit much? @albertgong1, do you have any feedback on this section?

File: src/nearest_neighbors/datasets/heartsteps/loader.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+                            final_M[user - 1, day - 1, slot - 1] = measurements
+                        else:
+                            m_pad = np.pad(
+                                measurements,
+                                (0, num_measurements - len(measurements)),
+                                constant_values=np.nan,
+                            )
+                            final_M[user - 1, day - 1, slot - 1] = m_pad
+                    except KeyError:
+
```
Comment (aashish-khub) [inline src/nearest_neighbors/datasets/heartsteps/loader.py:122]: rm blank line after except KeyError:

File: src/nearest_neighbors/datasets/heartsteps/loader.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+    def _get_mode(self, x: pd.Series) -> pd.Series:
+        if len(pd.Series.mode(x) > 1):
+            return pd.Series.mode(x, dropna=False)[0]
+        else:
+            return pd.Series.mode(x, dropna=False)
+
+
+    def _reind_id(self, df_u: pd.DataFrame) -> pd.DataFrame:
+        """Function to reindex the data to include all time points for each user."""
+        rng = pd.date_range(
```
Comment (aashish-khub) [inline src/nearest_neighbors/datasets/heartsteps/loader.py:141]: please use better variable names. "RNG" traditionally stands for random number generator even though i see it's clearly a range here

File: src/nearest_neighbors/datasets/heartsteps/loader.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+        )
+        df_sugg_sel = df_sugg_sel.dropna(
+            subset=["sugg.decision.utime", "sugg.select.utime", "user.index"]
+        )
+
+        # group the step data by five minute intervals
+        df_5min = (
+            df_steps.groupby(
+                [
+                    pd.Grouper(freq="5min", level="steps.utime", label="right"),
```
Comment (aashish-khub) [inline src/nearest_neighbors/datasets/heartsteps/loader.py:234]: 5min here is a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)). There's no reason this code should work for five minute intervals only. if changing this is involved please put in a #TODO and sign with your name in the comment & this can be addressed later—seems low priority but still must be done

File: src/nearest_neighbors/datasets/heartsteps/loader.py [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+            df_merged["decision.index.nogap"].isna(), np.nan, df_merged["sugg.select.slot"]
+        )
+
+        # get 12 rows after each notification period (1 hour of observations)
+        unique_users = df_merged["user.index"].unique()
+        result_dfs = []
+
+        for user_idx in unique_users:
+            df_user = df_merged[df_merged["user.index"] == user_idx]
+            df_user_range = self._take_range(df_user, 12)
```
Comment (aashish-khub) [inline src/nearest_neighbors/datasets/heartsteps/loader.py:288]: (this should also change once we get rid of the magic number)

File: src/nearest_neighbors/datasets/heartsteps/loader.py [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+            user_study_day = self._study_day(df_user)
+            study_day_results.append(user_study_day)
+
+        df_study_day = pd.concat(study_day_results)
+
+        # create unique index
+        df_final = df_study_day.set_index(["user.index", "study_day", "new_slot"])
+
+        # transform into 4d tensor + mask
+        Data, Mask = self._transform_dnn(df_final)
```
Comment (aashish-khub) [inline src/nearest_neighbors/datasets/heartsteps/loader.py:325]: unnecessarily capitalized variable name

File: pyproject.toml [hunk 13]
```diff
@@ -124,5 +126,5 @@ lint.unfixable = []
 [tool.pyright]
 venv = ".venv"
 venvPath = "."
-include = ["src"]
+include = ["src", "examples"]
```
[the remark this answers is not in the document]
Comment (calebchin) [reply pyproject.toml:129]: Didn't remove this from earlier -> when I had the data loader structure in examples this was needed for the imports.  Definitely can be removed now.

File: .gitignore [hunk 14]
```diff
@@ -161,4 +161,7 @@ cython_debug/
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
 .DS_Store
-.vscode/
\ No newline at end of file
+.vscode/
+
+*data/
+*out*/
```
[the remark this answers is not in the document]
Comment (calebchin) [reply .gitignore:167]: I think this is legacy from a previous commit where the standard output from the heartsteps processing would be data and out -> will remove this.

File: src/nearest_neighbors/dataloader_base.py (hunk 1 above)
Comment (calebchin) [reply src/nearest_neighbors/dataloader_base.py:10]: Got it, will fix

File: src/nearest_neighbors/datasets/heartsteps/loader.py (hunk 2 above)
Comment (calebchin) [reply src/nearest_neighbors/datasets/heartsteps/loader.py:26]: My intuition is that the processing step for scalar case is unique as some datasets may be more suited for the scalar case meaning that the distributional case (i.e. the process_data_distribution) will be the shorter one. For HeartSteps, the current preprocessing is meant for distributional, so converting to scalar is just a matter of aggregation. For another dataset, it might be the reverse (i.e. we expand a 2d matrix to 4d to mimic distributional case with 1 sample and dimensionality of 1).

File: src/nearest_neighbors/datasets/heartsteps/loader.py (hunk 3 above)
Comment (calebchin) [reply src/nearest_neighbors/datasets/heartsteps/loader.py:29]: Yeah this is just a bad habit from working with another codebase -> will remove

File: src/nearest_neighbors/datasets/heartsteps/loader.py (hunk 4 above)
Comment (aashish-khub) [reply src/nearest_neighbors/datasets/heartsteps/loader.py:46]: "this" being `Data.astype(object)`

Comment (calebchin) [reply src/nearest_neighbors/datasets/heartsteps/loader.py:46]: The data from the distributional processing is 4d (N x T x n x d) by convention so the aggregation over the axis 2 brings it to (N x T x d) with d = 1. So to get it to 2D, I squeeze to N x T. The conversion to object isn't pretty, but it is to align with the idea of the all NN input data being a 2D array of type object (if this is still the convention).

File: src/nearest_neighbors/datasets/heartsteps/loader.py (hunk 5 above)
Comment (calebchin) [reply src/nearest_neighbors/datasets/heartsteps/loader.py:60]: Sure, will change

File: src/nearest_neighbors/datasets/heartsteps/loader.py (hunk 6 above)
Comment (calebchin) [reply src/nearest_neighbors/datasets/heartsteps/loader.py:55]: The process_data_distribution method calls a helper (_proc_dist_data) which is very heartsteps specific.

File: .gitignore (hunk 14 above)
[the remark this answers is not in the document]
Comment (albertgong1) [reply .gitignore:167]: Yeah I added these. You can remove.

File: src/nearest_neighbors/datasets/heartsteps/loader.py (hunk 2 above)
Comment (calebchin) [reply src/nearest_neighbors/datasets/heartsteps/loader.py:26]: Will document this in docstring

**Review by aashish-khub — APPROVED** [review]
LGTM
